### PR TITLE
internal/v4: do not create new revision if content matches latest

### PR DESCRIPTION
**Pull request by rogpeppe**
When the content is the same as the latest version, there's no need to
bump the revision number. We needed to add another bundle to
test this, hence the change to the charm.v3 dependency.

### CODE REVIEW

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results: http://ci.jujugui.org:8080/job/charmstore/282/

File: internal/v4/archive.go [hunk 1]
```diff
@@ -106,6 +106,18 @@ func (h *Handler) servePostArchive(id *charm.Reference, w http.ResponseWriter, r
 		return badRequestf(nil, "Content-Length not specified")
 	}
 
+	oldId, oldHash, err := h.hashOfLatestRevision(id)
+	if err != nil && errgo.Cause(err) != params.ErrNotFound {
+		return errgo.Notef(err, "cannot get hash of latest revision")
+	}
+	if oldHash == hash {
```
Comment (frankban) [inline internal/v4/archive.go:113]: Very nice.

File: internal/charmstore/store.go [hunk 2]
```diff
@@ -143,21 +143,52 @@ func (s *Store) AddCharm(url *charm.Reference, c charm.Charm, blobName, blobHash
...
+			docs[last] = doc
+			last++
 		}
 	}
+	return docs[0:last], nil
+}
+
+// ExpandURL returns all the URLs that the given URL may refer to.
+func (s *Store) ExpandURL(url *charm.Reference) ([]*charm.Reference, error) {
+	entities, err := s.FindEntities(url, "_id")
```
Comment (frankban) [inline internal/charmstore/store.go:184]: Isn't the _id always returned, do we need to explicitly pass it here?

Comment (frankban) [inline internal/charmstore/store.go:184]: Ok, I presume we pass it to avoid all fields to be retrieved.

File: internal/v4/archive.go [hunk 3]
```diff
@@ -163,6 +175,23 @@ func (h *Handler) servePostArchive(id *charm.Reference, w http.ResponseWriter, r
 	})
 }
 
+func (h *Handler) hashOfLatestRevision(id *charm.Reference) (*charm.Reference, string, error) {
```
Comment (frankban) [inline internal/v4/archive.go]: latestRevisionInfo? It doesn't return just the hash. [shrug]

File: internal/v4/archive.go [hunk 4]
```diff
@@ -163,6 +175,23 @@ func (h *Handler) servePostArchive(id *charm.Reference, w http.ResponseWriter, r
...
+	entities, err := h.store.FindEntities(id)
```
Comment (frankban) [inline internal/v4/archive.go]: h.store.FindEntities(id, "url", "blobhash")? [shrug]

**Comment by frankban** [comment]
Very nice improvement Roger.
:+1: thank you!

File: internal/v4/archive.go (hunk 3 above)
Comment (rogpeppe) [inline internal/v4/archive.go]: done

File: internal/v4/archive.go (hunk 4 above)
Comment (rogpeppe) [inline internal/v4/archive.go]: good point - that use case was actually my motivation for including that feature of FindEntities!
done.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results: http://ci.jujugui.org:8080/job/charmstore/283/

File: internal/v4/archive.go [hunk 5]
```diff
@@ -106,6 +106,18 @@ func (h *Handler) servePostArchive(id *charm.Reference, w http.ResponseWriter, r
...
+	if err != nil && errgo.Cause(err) != params.ErrNotFound {
+		return errgo.Notef(err, "cannot get hash of latest revision")
+	}
+	if oldHash == hash {
+		// The hash matches the hash of the latest revision, so
+		// no need to upload anything.
+		return router.WriteJSON(w, http.StatusOK, &params.ArchivePostResponse{
+			Id: oldId,
+		})
+	}
```
Comment (urosj) [inline internal/v4/archive.go:119]: Nice!

**Comment by urosj** [comment]
:+1: Thank you.

**Comment by rogpeppe** [comment]
:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/charmstore-merge

**Pull request merged**